### ISSUE TITLE
the speed is not fast?

### ISSUE THREAD

**Issue by henbucuoshanghai** [ppogg/shufflev2-yolov5#22]
i test the pth in this git on my video,
cpu is 0.1s   not vert fast  why?
![Screenshot from 2021-08-31 10-22-51](https://user-images.githubusercontent.com/37647094/131431446-744b7ec3-c6f7-4da2-9cd6-e6bafa650208.png)

**Comment by henbucuoshanghai** [comment]
cpu

**Comment by ppogg** [comment]
Yeah, as you see, your test speed is about the same as mine

![image](https://user-images.githubusercontent.com/82716366/131431935-a4ea712c-16c4-4e6f-b68a-b6b4ac099add.png)

The backbone uses shufflenet. This kind of network mainly serves chips or end-side devices. Such as mobile phones, monitoring equipment, etc. More similar development boards are like to Raspberry Pi.

If you are only running on the cpu, it may be faster, and it is not practical at all on the gpu.

**Comment by henbucuoshanghai** [comment]
i gusee this mdoel on my phone is 15fps??maybe??
but i want to 35fps or 40fps on my phone,how to change the yolov5 model????

thank you

**Comment by ppogg** [comment]
> 我猜我手机上的这个 mdoel 是 15fps？？也许？？
> 但我想在我的手机上 35fps 或 40fps，如何更改 yolov5 模型？？？？
> 
> 谢谢你

这是在我手机适配的Android Demo,你可以下载这个[apk](https://github.com/ppogg/shufflev2-yolov5/blob/master/ncnn_Android/com.tencent.yolov5-release.apk) 进行安装,

![image](https://user-images.githubusercontent.com/82716366/131436328-7fecf61f-80b4-4f52-bf51-e20111cba914.png)

我的手机配置是redmi k30,处理器为骁龙730G,是一部两年的老手机,也是比较低端配置的手机,在开启检测一分钟后,稳定帧率为18FPS+

**Comment by henbucuoshanghai** [comment]
热心的回复。是的 差不多18fps的样子
但我想在我的手机上 35fps 或 40fps，如何更改 yolov5 模型？？？？

**Comment by ppogg** [comment]
> 热心的回复。是的 差不多18fps的样子 但我想在我的手机上 35fps 或 40fps，如何更改 yolov5 模型？？？？

好的,我想问下你的手机配置?

**Comment by henbucuoshanghai** [comment]
普通手机&nbsp; 小米789系列之类&nbsp; &nbsp;

**Comment by ppogg** [comment]
嗯嗯,收到了朋友,以下是回复:

首先一个,得看你手机支不支持超频,如果你手机摄像头只能支持每秒采集30fps,那么再快的模型在你手机运行也是30帧的,而且老手机性能很差的,你不能指望一千左右的手机能给你带来好几千的体验感(小米8和9可能还好,应该可以稳定22-25fps左右)

其他因素,单纯使用yolov5ss系列估计很难达到30fps,因为你要知道实际用的时候不止是向前推理,还有帧流预处理和后处理,这俩部分占用的时间在我的手机上(红米k30)大概是15ms左右(后处理还包含letter box),除非你使用的是华为p30这种处理器比较强大的手机

另外,yolov5ss的beachmark在红米k30测试过,向前推理一帧是25-30ms,华为p30没测过,估计一帧是10-15ms左右

如果你需要再快,既定你的检测类别不超过5类,有两种选择:
①把C3 block替换成depthwise:
```
class DWConvblock(nn.Module):
    "Depthwise conv + Pointwise conv"

    def __init__(self, in_channels, out_channels):
        super(DWConvblock, self).__init__()
        self.conv1 = nn.Conv2d(in_channels, in_channels, kernel_size=5, stride=1, padding=2, groups=in_channels,
                               bias=False)
        self.bn1 = nn.BatchNorm2d(in_channels)
        self.conv2 = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=1, padding=0, bias=False)
        self.bn2 = nn.BatchNorm2d(out_channels)

    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = self.bn2(x)
        return x
```
此处可以加速5%,但是会掉4个点左右

②把模型中的cat换成add,减少冗余参数:
```
class ADD(nn.Module):
    # Concatenate a list of tensors along dimension
    def __init__(self, alpha=1):
        super(ADD, self).__init__()
        self.a = alpha

    def forward(self, x):
        return torch.add(x, self.a)
```
此处还没测试过

③后期开源yolov5ss的预训练权重,大约可以加速25-35%,但是map仅有24,看你选择
![image](https://user-images.githubusercontent.com/82716366/131453249-fe55eff4-4dba-4c0b-8ddc-2918a9ec85d5.png)

**Comment by henbucuoshanghai** [comment]
热心有用的回复，感激感激！！
但是yolo5已经改进很多，shuflnet轻量，mobinet轻量，但是性能还是上不去。
那请问除了魔改yolo5，别的有专门做手机端的部署吗？出名的那种？50fps那种？

**Comment by ppogg** [comment]
是的,还是挺多的,这里推荐两个,一个是nanodet(在我的手机稳定22fps+),一个是yolo fastest(在我的手机稳定24fps+),这两个网络是很优秀的,,虽然在map@0.5上会略低于yolov5ss,但我认为可以忽略不记的,毕竟实际检测不会涉及到80类这么多

**Comment by henbucuoshanghai** [comment]
你好&nbsp; 我就一类
但是nanodet他上面写着手机上60fps之类的？怎么才22？
他们比yolo3 tiny 如何？

**Comment by ppogg** [comment]
nanodet的作者使用的是华为p30测的beachmark,红米的处理器很差的,两者差个两三倍我觉得挺正常

yolov3 tiny和v4 tiny已经是过去式了,flops厚重,实际推理较慢,大部分落地还是会考虑更轻量的模型,但也不能说他们不好,其实对于一些芯片,就不见得轻量的模型比tiny系列好,这里推荐一篇博客,你可以看看:https://zhuanlan.zhihu.com/p/340959058

实际落地需要考虑cache,缓存和访存等很多东西,需要见缝插针,不断试错

**Comment by henbucuoshanghai** [comment]
博客你写的？
好专业。。刚接触tnn手机端。大佬研究的好透。。
那同一个手机里，nanodet肯定比yolo34 tiny好多了是吗？


另：git提供的pth测试视频可以，但是我转成onnx，tnn后当初的后处理代码，跑完没结果。。没任何box，怎么回事。

**Comment by ppogg** [comment]
博客是qiuqiu大佬写的，他是这方面的行家

你说的pth是哪个文件，我这边没有提供pth文件额

**Comment by henbucuoshanghai** [comment]
weights文件夹下 yolo5ss.pt

**Comment by henbucuoshanghai** [comment]
感谢昨天的系列回答，很有用，感激！最后几个问题
1：你的博客是女生，不会吧女生？
2：刚学tnn请问yolofastest哪里有tnn魔板呢？读取模型，后处理那种？
3：tnn是ncnn的加强版？就调用模型命令不一样，一个算法&nbsp; 前后处理可以参考ncnn的后处理？
最后&nbsp; 有什么微信群qq群什么的博客什么的学习圈子呢？tnn这一块，手机部署检测，姿态这一块。


感谢感谢！！！！

**Comment by chenweifu2008** [comment]
大佬请问一下，我下载了最新版的ncnn，在电脑上安装和编译也都过了没啥问题，跑了你的yolov5ss.cpp 和你提供的ncnn模型，为什么我这边速度很慢大概一两秒才能出一张结果，而且那张person的图片竟然有一个误检。我跑pytorch模型速度(cpu)就相对正常。
请问您ncnn部署测试是怎么做的呢？是因为ncnn在电脑上跑本来就很慢吗也不太可能一两秒吧？谢谢

**Comment by ppogg** [comment]
> 大佬请一下，我下载了最新版的ncnn，在电脑上安装和编译也都过了没啥问题，跑了你的yolov5s.cpp和你提供的ncnn模型，为什么我这个速度很慢大概一我跑pytorch模型速度(cpu)就相对。
> 请问您ncnn开发测试是怎么做的呢？是ncnn在电脑上跑本来就很慢吗也不太可能一两秒吧？谢谢

麻烦另开一个issue，顺便告诉我在什么平台跑的，贴一下bin和param文件

**Comment by ppogg** [comment]
> 感谢昨天的系列回答，很有用，感激！最后几个问题 1：你的博客是女生，不会吧女生？ 2：刚学tnn请问yolofastest哪里有tnn魔板呢？读取模型，后处理那种？ 3：tnn是ncnn的加强版？就调用模型命令不一样，一个算法&nbsp; 前后处理可以参考ncnn的后处理？ 最后&nbsp; 有什么微信群qq群什么的博客什么的学习圈子呢？tnn这一块，手机部署检测，姿态这一块。 感谢感谢！！！！

1、男生，为什么会这么问。。。
2、这个需要你去yolo fastest那提下issue
3、tnn和ncnn是同一家公司，模型部署主要还是前处理，模型加载，后处理，但后处理在不同平台上可能略有不同
4、估计没有，网上的资料都很零散，但是你可以加ncnn的群问问，里面很多人会mnn和tnn

**Comment by henbucuoshanghai** [comment]
你好 你的博客显示女性明白&nbsp;
那请问手机端这些快的模型，score阈值很小对吗，0.5就很高了？

**Comment by ppogg** [comment]
> 那请问手机端这些快的模型，score阈值很小对吗，0.5就很高了？

不一定的，yolov4-tiny会高些，但是我这边使用了yolov5的正负样本匹配机制，分值都被压得特别低

**Comment by henbucuoshanghai** [comment]
你好，最近二个月在做姿态，请问2d姿态有很好的手机版开源吗？轻量化训练2d姿态？请问那个github成熟呢？感谢
